aea/aea.py

```diff
@@ -239,7 +239,10 @@ def _handle(self, envelope: Envelope) -> None:
             return
 
         try:
-            msg = protocol.serializer.decode(envelope.message)
+            if isinstance(envelope.message, Message):
+                msg = envelope.message
+            else:
+                msg = protocol.serializer.decode(envelope.message)
```

Comment:
Incoming messages from connecition might or might not be serialized

Reply:
Small detail: I'd prefer the other way around; check for `bytes` first; then in the `else` branch save in `msg` whatever is there. `Message` can change in the future, `bytes` won't.

Reply:
> Incoming messages from connection might or might not be serialized

Why not giving the connection objects access to the serializers? It would give us "design symmetry": for outgoing envelopes we will have `isinstance(envelope.message, Message)`, and for incoming messages we will have the same, rather than "maybe bytes".

Reply:
It also connects to the "restricted_protocols" and "excluded_protocols" fields: the connection will only have/won't have access (respectively) to those protocols.

I know, it's not an easy one... Maybe in the next releases.

Reply:
If I invert the check, then mypy will complain due to it not knowing that it is of `Message` type. Giving the connections access to the serialisers means a significant rewrite. We would have to provide each connection with the dictionary of all serialisers which means resources bleed into the multiplexer. I tried and could not find a clean enough approach. I think this approach, although asymmetric is neater; as all the connection does is deal with an envelope on a per instance basis.